### ISSUE TITLE
Regression: when debugger encounters a not-unhandled exception, it stops and shows "Dns.cs file not found" (Dns being the BCL class used)

### ISSUE THREAD

**Issue by knocte** [mono/monodevelop#8552]
I believe this is a regression because I haven't seen this behaviour in previous versions of VisualStudio for Mac (now using 8.2.4; not sure the exact version when this started happening, possibly when Mono 6.0 was bundled).

Steps to reproduce:
1. Open VS4Mac, create new project/solution, Console type, C#.
2. Use this simple snippet:

```
using System;
using System.Linq;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Threading.Tasks;

namespace TestResolveCSharpConsole
{
    class MainClass
    {
        public static void Main(string[] args)
        {
            try
            {
                var results = ResolveAsync().GetAwaiter().GetResult();
            }
            catch (SocketException sockEx)
            {
                Console.WriteLine("correctly handled");
            }
        }

        public static async Task<List<string>> ResolveAsync()
        {
            var addresses = await System.Net.Dns.GetHostAddressesAsync("foo.bar.baz");
            return new List<string>(addresses.Select(addr => addr.ToString()));
        }
    }
}
```

3. Compile it. (If you run it outside VS4Mac, result should just be the printing of the string "correctly handled.)
4. Run it with the debugger inside VS4Mac.

Expected results:
Should just run it and not stop, finish execution at the end like it's the case when being run outside of VS4Mac.

Current results:
Debugger seems to bomb when a `SocketException` is thrown, even if it's properly handled with a try-catch (this is the first bug). Then it tries to open the source code of the BCL where this exception happened (Dns.cs), telling the weird error "Source Not Found" to the user.

See screenshot:
<img width="1270" alt="Screenshot 2019-08-28 at 3 56 16 PM" src="https://user-images.githubusercontent.com/331303/63837030-46735700-c9ad-11e9-9d23-c24a2e133601.png">

**Comment by jstedfast** [comment]
Pretty sure this is caused by the new "SourceLink" feature added recently by @nosami :-\

**Comment by nosami** [comment]
Haven't attempted to repro this yet, but it looks like 2 different bugs to me. The 1st bug is that the debugger shouldn't pause at all. The 2nd bug is that if the debugger does break for this reason, then SourceLink should have correctly found the Dns.cs file to download (provided this feature is even turned on, the bug report doesn't mention this)

**Comment by nosami** [comment]
@jeff with SourceLink enabled, the debugger correctly shows the location of the code that shows the exception

<img width="1044" alt="image" src="https://user-images.githubusercontent.com/667194/65629503-5e99be80-dfcb-11e9-9799-dfc52cfff590.png">

The issue here though is that the debugger shouldn't break at all, not SourceLink AFAICT

**Comment by nosami** [comment]
The same bug happens with VSMac 8.2.3.5 before SourceLink was even introduced + Mono 6.4.0.198. Pretty sure that this is a regression in Mono.

**Comment by nosami** [comment]
@thaystg Please could you look at this and see if this is a bug in Mono? Thanks!

**Comment by knocte** [comment]
>  (provided this feature is even turned on, the bug report doesn't mention this)

It doesn't mention this because this bug was reported when stable channel only had the 8.2.6 version of VS4Mac. It's 8.3.x the one that introduces SourceLink, right? Because I haven't upgraded yet to 8.3.x and I don't see the setting in the debugger preferences page, so SourceLink must be disabled.

**Comment by knocte** [comment]
> but it looks like 2 different bugs to me. The 1st bug is that the debugger shouldn't pause at all. The 2nd bug is that if the debugger does break for this reason, then SourceLink should

But what if there's no SourceLink(8.2.x) or SourceLink is disabled(8.3.x)? The risk of fixing the 1st bug first is that then it will hide the 2nd bug under the rug (this wouldn't mean that the 2nd bug is fixed, just less likely to happen). IMHO the 2nd bug should be fixed first.

**Comment by nosami** [comment]
@knocte there isn't a 2nd bug after all, only a first one :)

**Comment by knocte** [comment]
> @knocte there isn't a 2nd bug after all, only a first one :) SourceLink isn't connected to this issue. I repro'd using VSMac 8.1 / Mono 6.4.0.198 too

I'm guessing that what you mean is that it's not SourceLink related, sure, but the 2nd bug still exists. Fixing only the 1st one might just "hide" it for this scenario, but could resurface in another one (e.g. if user configures debugger to stop in all exceptions, even handled ones?).

**Comment by nosami** [comment]
No. What I meant is that there isn't a 2nd bug.

The page that you are seeing is what happens _by design_ when there is no source code for the current location irrespective of SourceLink.

**Comment by knocte** [comment]
> when there is no source code for the current location

And shouldn't the source code of Mono be shipped by VS4Mac so that a dev can step into the BCL?

**Comment by nosami** [comment]
> And shouldn't the source code of Mono be shipped by VS4Mac so that a dev can step into the BCL?

No. That was the _whole point_ of adding SourceLink - so that we can stop adding source code in the various Mono pkgs.

**Comment by nosami** [comment]
The bug here is that the debugger stops on handled exceptions when it should not.

**Comment by knocte** [comment]
>  so that we can stop adding source code in the various Mono pkgs.

But Mono source code is not available through nuget. I thought SourceLink was integration with nuget? Can then VS4Mac8.3 show Dns.cs from Mono out of the box?

**Comment by nosami** [comment]
No. SourceLink is a technology that isn't connected to NuGet but is used by many NuGet packages.

> Can then VS4Mac8.3 show Dns.cs from Mono out of the box?

Yes, with these settings

<img width="549" alt="image" src="https://user-images.githubusercontent.com/667194/65678936-1b336480-e04c-11e9-9c2e-a974eb1573cc.png">

**Comment by knocte** [comment]
Cool, understood!

@thaystg let me know if you need any help reproducing it. You might want to add more parallelisation to the testcase (as the snippet I posted was just my effort to convert the bug to a minimal repro; but I initially reproduced it with many more async parallel jobs running simultaneously, not just one).
